sycl/include/CL/sycl/detail/context_impl.hpp

```diff
@@ -58,14 +58,25 @@ class context_impl {
   RT::PiContext &getHandleRef();
   const RT::PiContext &getHandleRef() const;
 
-  std::map<OSModuleHandle, RT::PiProgram> &getCachedPrograms() {
+  std::map<OSModuleHandle, RT::PiProgram> &acquireCachedPrograms() {
+    m_ProgramCacheMutex.lock();
```

Comment:
You should not use explicit `lock()`/`unlock() but use the RAII interface instead.
But actually your `acquireCachedKernels()`/`releaseCachedKernels()` looks very C++-unfriendly.
What happens for example if you have an exception happening in the middle? :-(
Perhaps https://github.com/triSYCL/triSYCL/blob/master/include/triSYCL/detail/cache.hpp can help.

Reply:
Employed RAII along with acquire/release.
Acquire method returns a class instance (locked) with reference to cache and a lock for the cache.
Release will happen upon destruction of the locked.